### PR TITLE
Adds new `did_you_mean` algorithm to enhance error messages.

### LINKED ISSUE #2682: proposal: do not show "did you mean X" if the edit distance is too great
Hello, Gleam developers

I was just trying out Gleam at https://tour.gleam.run/ and saw this confusing message.

```
error: Unknown variable
  ┌─ /src/main.gleam:4:3
  │
4 │   println("My lucky number is:")
  │   ^^^^^^^ Did you mean `Nil`?

The name `println` is not in scope here.
```

It took me awhile to figure out what "Did you mean `Nil`?" meant.

To my memory, Rust won't show the same error message if the identifier is too far from any name in scope.

Here is the error message without the confusing part:

```
error: Unknown variable
  ┌─ /src/main.gleam:4:3
  │
4 │   println("My lucky number is:")
  │   ^^^^^^^

The name `println` is not in scope here.
```

which i argue is better.

### PR DESCRIPTION
**Pull request by tubedude**
This will close #2682 , but also contains the changes for #2791 

This changes about 15 snapshots with what I consider more relevant error messages. But should be doubled checked.

### CODE REVIEW

**Comment by lpil** [comment]
Hello! Could you remove the commits from the other PR please 🙏

**Comment by lpil** [comment]
Could you remove the duplicated commits and then rebase to resolve the conflicts please? Thank you

**Comment by lpil** [comment]
Hello! Could you revert [8d8b274](https://github.com/gleam-lang/gleam/pull/2798/commits/8d8b274c19b58838adca9b21bea846cd7fb0a945) please to maintain the existing behaviour.

**Comment by tubedude** [comment]
Louis! sorry the mess on the commit. I rebased to main, and applied the change on the `error.rs` (and affected snapshots).
I've maintain the _one option exception_ as discussed, and solved all _clippy_ issues, and ran rustfmt

**Review by lpil — COMMENTED** [review]
Hmm, the linter should have banned that use of `unwrap`. Not sure why it didn't. Could you swap those for `expect` please, to make the panic messages more useful for debugging.

Otherwise, looks fantastic! We'll need a changelog entry too. Thank you

**Comment by lpil** [comment]
Could you revert the unrelated changes to the changelog please, we do one change per PR to make it easier to review and merge.

**Comment by tubedude** [comment]
Done!
- Reverted the changelog reorg;
- updated branch
- Added entry related to this PR to changelog.

File: CHANGELOG.md [hunk 1]
```diff
@@ -50,6 +50,8 @@
 - Improved error messages when failing to parse a series of things.
 - A warning is now raised for unused binary operations, records, record access
   and record updates.
+- Improve error messages with a new `did_you_mean` algorithm; and added note
+  when trying to access custom types with multi-variant.
```
Comment (lpil) [inline CHANGELOG.md:54]: This isn't right- this feature isn't in this PR.

Could you remove the reference to did_you_mean too as that's an implementation detail of the compiler and the changelog is written for users. We could just say that suggestions have been improved.

**Review by lpil — COMMENTED** [review]
Fab! Could you rebase to remove the merge commits please. Thank you.

File: CHANGELOG.md (hunk 1 above)
Comment (tubedude) [reply CHANGELOG.md]: Done. And thanks for the patience.
```
- Improved recommendations on error messages.
```

**Review by lpil — APPROVED** [review]
Wonderful work! Thank you so much!

And thank you for being patient as I slowly get through my inbox 😁

**Pull request merged**